### ISSUE TITLE
Connection Error certificate verification failed

### ISSUE THREAD

**Issue by satra** [jtriley/StarCluster#267]
getting a connection error with the latest git version of starcluster. 

```
$ starcluster createkey myawskey -o ~/.ssh/myawskey.rsa
StarCluster - (http://star.mit.edu/cluster) (v. 0.9999)
Software Tools for Academics and Researchers (STAR)
Please submit bug reports to starcluster@mit.edu

!!! ERROR - Connection error:
Traceback (most recent call last):
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/starcluster/cli.py", line 257, in main
    sc.execute(args)
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/starcluster/commands/createkey.py", line 27, in execute
    kp = self.ec2.create_keypair(name, output_file=ofile)
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/starcluster/awsutils.py", line 426, in create_keypair
    kp = self.conn.create_key_pair(name)
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/ec2/connection.py", line 2176, in create_key_pair
    return self.get_object('CreateKeyPair', params, KeyPair, verb='POST')
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/connection.py", line 1063, in get_object
    response = self.make_request(action, params, path, verb)
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/connection.py", line 989, in make_request
    return self._mexe(http_request)
  File "/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/connection.py", line 908, in _mexe
    raise e
SSLError: [Errno 1] _ssl.c:504: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
!!! ERROR - Check your internet connection?
```

**Comment by jtriley** [comment]
I see this from time to time myself - doing a bit more research I found this:

http://stackoverflow.com/questions/13321302/python-foursquare-ssl3-certificate-verify-failed

Turns out @ogrisel commented that it worked for boto as well. Perhaps boto simply needs a newer cacerts? What version of boto are you using btw?

/cc @garnaat

**Comment by ogrisel** [comment]
Yes I indeed had to update my cacerts.txt in the boto folder to make StarCluster work again. In my case I currently have boto 2.9.6.

**Comment by jtriley** [comment]
@ogrisel It wouldn't work _at all_ without an update to cacerts.txt? For me this only happens once in a blue moon...

**Comment by satra** [comment]
thanks. that works like a charm - i was using boto 2.9.6 as well. 

for me every call on os x box was stalling, prior to the change.

**Comment by ogrisel** [comment]
@jtriley yes I could no longer use boto and StarCluster prior to updating the certificates.

**Comment by jtriley** [comment]
Wow, that's a major issue. I've pinged folks on #boto IRC - we'll see what they say. My guess is we need a PR for boto to update the cacerts file but there may be a larger issue here with system certificates.

**Comment by jtriley** [comment]
@ogrisel For the sake of fixing the bigger issue would you mind reproducing this issue and then running:

```
$ starcluster shell
[1]> logger.configure_boto_logging()
[2]> ec2.list_instances() 
```

Then look for SSL lines in `$HOME/.starcluster/logs/aws-debug.log` like:

```
PID: 12081 DEB [20130701-18:26:22.454] boto: wrapping ssl socket; CA certificate file=/home/jtriley/.virtualenvs/starcluster/lib/python2.6/site-packages/boto-2.9.4-py2.6.egg/boto/cacerts/cacerts.txt
PID: 12081 DEB [20130701-18:26:22.861] boto: validating server certificate: hostname=ec2.us-east-1.amazonaws.com, certificate hosts=[u'ec2.us-east-1.amazonaws.com']
```

As you can see my cacerts.txt within boto works fine but I'm guessing for you there's an error in the logs that should help get to the bottom of this. Please paste any relevant lines from that file but be careful not to copy any sensitive info like urls that have AWS_ACCESS_KEY in them, etc.

Also, I found a similar (closed) issue boto/boto#1535

**Comment by jtriley** [comment]
@ogrisel @satra What OS version are you running into this issue on? Do you have this issue on other machines with a different OS? The crash report has the exact details if you still have it. Thanks.

**Comment by satra** [comment]
here you go and i'm running on osx 10.8.4 and this repeats a few times.

```
PID: 63819 DEB [20130701-18:37:54.513] boto: establishing HTTPS connection: host=ec2.us-east-1.amazonaws.com, kwargs={'timeout': 70}
PID: 63819 DEB [20130701-18:37:55.255] boto: Token: None
PID: 63819 DEB [20130701-18:37:55.255] boto: using _calc_signature_2
PID: 63819 DEB [20130701-18:37:55.255] boto: query string: AWSAccessKeyId=++++&Action=DescribeImages&Owner.1=++++&SignatureMethod=HmacSHA256&SignatureVersion=2&Timestamp=2013-07-01T22%3A37%3A55Z&Version=2013-02-01
PID: 63819 DEB [20130701-18:37:55.256] boto: string_to_sign: POST
ec2.us-east-1.amazonaws.com
/
AWSAccessKeyId=++++&Action=DescribeImages&Owner.1=++++&SignatureMethod=HmacSHA256&SignatureVersion=2&Timestamp=2013-07-01T22%3A37%3A55Z&Version=2013-02-01
PID: 63819 DEB [20130701-18:37:55.256] boto: len(b64)=44
PID: 63819 DEB [20130701-18:37:55.256] boto: base64 encoded digest: ++++
PID: 63819 DEB [20130701-18:37:55.256] boto: query_string: AWSAccessKeyId=++++&Action=DescribeImages&Owner.1=++++&SignatureMethod=HmacSHA256&SignatureVersion=2&Timestamp=2013-07-01T22%3A37%3A55Z&Version=2013-02-01 Signature: ++++
PID: 63819 DEB [20130701-18:37:55.330] boto: wrapping ssl socket; CA certificate file=/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/cacerts/cacerts.txt
PID: 63819 DEB [20130701-18:37:55.412] boto: encountered SSLError exception, reconnecting
```

**Comment by jtriley** [comment]
@satra Thanks unfortunately that doesnt tell us much. Also do you have a `$HOME/.boto` by chance?

**Comment by satra** [comment]
no `$HOME/.boto` but i can replicate the previous error in: boto/boto#1535 with either ec2 on east or west

```
=============================
Boto Version: 2.9.6
=============================

Testing Good Cert:/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/cacerts/new_cacerts.txt
==================
Connecting socket
Wrapping with SSL
Cert info: {'notAfter': 'Oct  7 23:59:59 2013 GMT', 'subjectAltName': (('DNS', 'ec2.us-east-1.amazonaws.com'), ('DNS', 'ec2.amazonaws.com'), ('DNS', 'us-east-1.ec2.amazonaws.com')), 'subject': ((('countryName', u'US'),), (('stateOrProvinceName', u'Washington'),), (('localityName', u'Seattle'),), (('organizationName', u'Amazon.com Inc.'),), (('commonName', u'ec2.us-east-1.amazonaws.com'),))}

Testing Bad Cert:/software/virtualenvs.EPD/7.3/devpype/lib/python2.7/site-packages/boto/cacerts/cacerts.txt
==================
Connecting socket
Wrapping with SSL
FAILED: [Errno 1] _ssl.c:504: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
```

**Comment by jtriley** [comment]
@ogrisel are you using EPD as well?

**Comment by ogrisel** [comment]
No I used python from homebrew and boto from pip and starcluster from git master.

**Comment by jtriley** [comment]
Just for the record: Now that boto/boto#1578 has been merged I will undo 6f26ab5 as soon as the next version of boto is released.
